docs/madr/decisions/056-reachable-svc-mesh-service.md

```diff
@@ -0,0 +1,239 @@
+
+## Considered Options
+
+* Support reachable and autoreachable services for `MeshService` and `MeshExternalService`
+
+## Decision Outcome
+
+Chosen option: Support reachable and autoreachable services for `MeshService` and `MeshExternalService`
+
+### Reachable services
+
+In the case of Kubernetes, we can introduce a new annotation that has a more structured model and better flexibility.
+
+```yaml
+kuma.io/reachable-backend-refs: |
+  refs:
+  - kind: MeshService
+    name: demo-app
+    namespace: kuma-demo # if missing, then my namespace
+    port: 5000
+  - kind: MeshService
+    name: redis
+    namespace: redis-system
+  - kind: MeshService
+    labels:
+      kuma.io/display-name: xyz
+      kuma.io/zone: east
+  - kind: MeshExternalService
+    name: demo-app
+    namespace: kuma-system
+  - kind: MeshExternalService
+    labels:
+      kuma.io/display-name: httpbin
+```
+
+This model is more extendable, and if we introduce a new kind in the future, we can easily extend it. It is also similar to targetRef. The disadventage might be it takes much more space, and might be a bit difficult to manage within an annotation.
+
+Universal needs another entry in a `Dataplane` object:
+
+```proto
+message TransparentProxying {
+...
+    message ReachableBackendRef {
+        // Type of the backend: MeshService or MeshExternalService
+        //  +required
+        string kind = 1;
+
+        // Name of the backend.
+        //  +optional
+        string name = 2;
+
+        // Namespace of the backend. Might be empty
+        //  +optional
+        string namespace = 3;
+
+        // Port of the backend.
+        //  +optional
+        google.protobuf.UInt32Value port = 4;
+
+        // Labels used to select backends
+        //  +optional
+        map<string, string> labels = 5;
+    }
+
+    // List of reachable backend refs via transparent proxy when running with 
+    // MeshExternalService and MeshService. Setting an explicit list
+    // can dramatically improve the performance of the mesh. If not specified,
+    // all services in the mesh are reachable.
+    repeated ReachableBackendRef reachable_backends_ref = 7;
+    }
+```
+
+Most of the fields are optional and may not be used simultaneously. We need to support the following cases:
+* When `name`, `namespace`, and `port` are defined(combination of them, all are not requried), you cannot use `labels`.
+* When `labels` are defined, you cannot use `name`, `namespace`, and `port`.
+
+Another thing that is required to do is add another fields to the `BackendRef` in Dataplane object: namespace and labels
+```proto
+      message BackendRef {
+        // Kind is a type of the object to target. Allowed: MeshService, MeshExternalService
+        string kind = 1;
+        // Name of the targeted object
+        string name = 2;
+        // Port of the targeted object. Required when kind is MeshService, MeshExternalService
+        uint32 port = 3;
+        // Namespace of the targeted object.
```

Comment:
I am not yet fully certain if this is necessary, as the name might already include this functionality. The name is a bit tricky because, once it's synced, it has a hash and namespace appended, which can cause a lot of confusion.

Reply:
I think this should align with the backendRef of routes ultimately, right?

Reply:
It should, but I feel it might not. Currently, backendRef in a dataplane's outbound has a name in the format `name-hash.namespace` (hash when synced), while backendRef in a route has `name-hash` with the namespace in a separate field. That's why I am considering doing the same for outbounds.